File: backend/fastapi-app/app/services/rbac_service.py
"""RBAC 关联业务 —— 对应 RbacService：用户-角色 / 角色-权限 分配与聚合查询。

聚合查询用原生 SQL（对应 RbacMapper.xml 的多表 join），
其余关联操作用 ORM（对应 JPA Repository）—— 双持久层协同，与 Java 版一致。
"""
from datetime import datetime

from sqlalchemy import text
from sqlalchemy.orm import Session

from .. import models
from ..schemas import UserPermissionVO
from .rbac_services import PermissionService, RoleService
from .user_service import BizError


class RbacService:
    def __init__(self, db: Session) -> None:
        self.db = db
        self.role_service = RoleService(db)
        self.permission_service = PermissionService(db)

    # ── 用户-角色 ──────────────────────────────────────────────

    def get_user_roles(self, user_id: int) -> list[models.Role]:
        if self.db.get(models.User, user_id) is None:
            raise BizError(f"用户不存在，ID: {user_id}")
        rows = (
            self.db.query(models.Role)
            .join(models.UserRole, models.UserRole.role_id == models.Role.id)
            .filter(models.UserRole.user_id == user_id)
            .order_by(models.Role.id)
            .all()
        )
        return rows

    def assign_role_to_user(self, user_id: int, role_id: int) -> models.Role:
        if self.db.get(models.User, user_id) is None:
            raise BizError(f"用户不存在，ID: {user_id}")
        role = self.role_service.find_by_id(role_id)
        exists = (
            self.db.query(models.UserRole)
            .filter(models.UserRole.user_id == user_id, models.UserRole.role_id == role_id)
            .first()
        )
        if not exists:  # 幂等：已分配则忽略
            self.db.add(models.UserRole(user_id=user_id, role_id=role_id))
            self.db.commit()
        return role

    def remove_role_from_user(self, user_id: int, role_id: int) -> None:
        if self.db.get(models.User, user_id) is None:
            raise BizError(f"用户不存在，ID: {user_id}")
        self.role_service.find_by_id(role_id)
        self.db.query(models.UserRole).filter(
            models.UserRole.user_id == user_id, models.UserRole.role_id == role_id
        ).delete()
        self.db.commit()

    # ── 角色-权限 ──────────────────────────────────────────────

    def get_role_permissions(self, role_id: int) -> list[models.Permission]:
        self.role_service.find_by_id(role_id)
        rows = (
            self.db.query(models.Permission)
            .join(models.RolePermission, models.RolePermission.permission_id == models.Permission.id)
            .filter(models.RolePermission.role_id == role_id)
            .order_by(models.Permission.id)
            .all()
        )
        return rows

    def assign_permission_to_role(self, role_id: int, permission_id: int) -> models.Permission:
        self.role_service.find_by_id(role_id)
        permission = self.permission_service.find_by_id(permission_id)
        exists = (
            self.db.query(models.RolePermission)
            .filter(
                models.RolePermission.role_id == role_id,
                models.RolePermission.permission_id == permission_id,
            )
            .first()
        )
        if not exists:  # 幂等
            self.db.add(models.RolePermission(role_id=role_id, permission_id=permission_id))
            self.db.commit()
        return permission

    def remove_permission_from_role(self, role_id: int, permission_id: int) -> None:
        self.role_service.find_by_id(role_id)
        self.permission_service.find_by_id(permission_id)
        self.db.query(models.RolePermission).filter(
            models.RolePermission.role_id == role_id,
            models.RolePermission.permission_id == permission_id,
        ).delete()
        self.db.commit()

    # ── 聚合查询（对应 RbacMapper.xml findUserPermissions） ───────

    def get_user_permissions(self, user_id: int) -> list[UserPermissionVO]:
        if self.db.get(models.User, user_id) is None:
            raise BizError(f"用户不存在，ID: {user_id}")
        sql = text(
            """
            SELECT p.id AS id,
                   p.code AS code,
                   p.name AS name,
                   p.description AS description,
                   r.code AS roleCode,
                   r.name AS roleName,
                   ur.created_at AS grantedAt
            FROM user_roles ur
                JOIN roles r ON ur.role_id = r.id
                JOIN role_permissions rp ON rp.role_id = r.id
                JOIN permissions p ON rp.permission_id = p.id
            WHERE ur.user_id = :userId
            ORDER BY p.id, r.id
            """
        )
        rows = self.db.execute(sql, {"userId": user_id}).all()
        return [
            UserPermissionVO(
                id=row.id,
                code=row.code,
                name=row.name,
                description=row.description,
                roleCode=row.roleCode,
                roleName=row.roleName,
                grantedAt=row.grantedAt,
            )
            for row in rows
        ]

File: spring-boot/src/main/java/com/example/demo/service/RbacService.java
package com.example.demo.service;

import com.example.demo.cache.CacheInvalidationService;
import com.example.demo.cache.CacheKeyFactory;
import com.example.demo.cache.MultiLevelCache;
import com.example.demo.mapper.RbacMapper;
import com.example.demo.pojo.entity.Permission;
import com.example.demo.pojo.entity.Role;
import com.example.demo.pojo.entity.RolePermission;
import com.example.demo.pojo.entity.UserRole;
import com.example.demo.pojo.vo.UserPermissionVO;
import com.example.demo.repository.PermissionRepository;
import com.example.demo.repository.RolePermissionRepository;
import com.example.demo.repository.RoleRepository;
import com.example.demo.repository.UserRepository;
import com.example.demo.repository.UserRoleRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

/**
 * RBAC 关联管理服务：用户-角色、角色-权限 的分配/移除，以及权限聚合查询。
 * <p>缓存策略：关联关系本身不直接缓存（写热点、一致性枢纽），
 * 缓存其“派生读模型”：user-roles / role-permissions / user-permissions；
 * 任何关系写入都在事务提交后失效对应派生 key，并由 Redis Pub/Sub 同步其他 JVM。
 */
@Service
@RequiredArgsConstructor
public class RbacService {

    public static final String CACHE_USER_ROLES = "user-roles";
    public static final String CACHE_ROLE_PERMISSIONS = "role-permissions";
    public static final String CACHE_USER_PERMISSIONS = "user-permissions";

    private final UserRepository userRepository;
    private final RoleRepository roleRepository;
    private final PermissionRepository permissionRepository;
    private final UserRoleRepository userRoleRepository;
    private final RolePermissionRepository rolePermissionRepository;
    private final RbacMapper rbacMapper;
    private final MultiLevelCache cache;
    private final CacheInvalidationService invalidation;

    // ── 用户-角色 ──────────────────────────────────────────────

    /** 查询用户拥有的角色列表（L1 → L2 → DB） */
    public List<Role> getUserRoles(Long userId) {
        requireUser(userId);
        return cache.getList(CACHE_USER_ROLES, CacheKeyFactory.id(userId), Role.class,
                () -> {
                    List<Long> roleIds = userRoleRepository.findByUserId(userId).stream()
                            .map(UserRole::getRoleId)
                            .collect(Collectors.toList());
                    return roleIds.isEmpty() ? List.of() : roleRepository.findAllById(roleIds);
                });
    }

    /** 给用户分配角色（幂等：已分配则忽略） */
    @Transactional
    public Role assignRoleToUser(Long userId, Long roleId) {
        requireUser(userId);
        Role role = requireRole(roleId);
        if (!userRoleRepository.existsByUserIdAndRoleId(userId, roleId)) {
            userRoleRepository.save(UserRole.builder().userId(userId).roleId(roleId).build());
        }
        invalidation.evictUserRoleLinks(userId);
        return role;
    }

    /** 移除用户的角色 */
    @Transactional
    public void removeRoleFromUser(Long userId, Long roleId) {
        requireUser(userId);
        requireRole(roleId);
        userRoleRepository.deleteByUserIdAndRoleId(userId, roleId);
        invalidation.evictUserRoleLinks(userId);
    }

    // ── 角色-权限 ──────────────────────────────────────────────

    /** 查询角色拥有的权限列表（L1 → L2 → DB） */
    public List<Permission> getRolePermissions(Long roleId) {
        requireRole(roleId);
        return cache.getList(CACHE_ROLE_PERMISSIONS, CacheKeyFactory.id(roleId), Permission.class,
                () -> {
                    List<Long> permissionIds = rolePermissionRepository.findByRoleId(roleId).stream()
                            .map(RolePermission::getPermissionId)
                            .collect(Collectors.toList());
                    return permissionIds.isEmpty() ? List.of() : permissionRepository.findAllById(permissionIds);
                });
    }

    /** 给角色分配权限（幂等：已分配则忽略） */
    @Transactional
    public Permission assignPermissionToRole(Long roleId, Long permissionId) {
        requireRole(roleId);
        Permission permission = requirePermission(permissionId);
        if (!rolePermissionRepository.existsByRoleIdAndPermissionId(roleId, permissionId)) {
            rolePermissionRepository.save(RolePermission.builder().roleId(roleId).permissionId(permissionId).build());
        }
        invalidation.evictRolePermissionLinks(roleId);
        return permission;
    }

    /** 移除角色的权限 */
    @Transactional
    public void removePermissionFromRole(Long roleId, Long permissionId) {
        requireRole(roleId);
        requirePermission(permissionId);
        rolePermissionRepository.deleteByRoleIdAndPermissionId(roleId, permissionId);
        invalidation.evictRolePermissionLinks(roleId);
    }

    // ── 聚合查询 ───────────────────────────────────────────────

    /** 查询用户拥有的全部权限（MyBatis 多表 join，含角色来源；L1 → L2 → DB） */
    public List<UserPermissionVO> getUserPermissions(Long userId) {
        requireUser(userId);
        return cache.getList(CACHE_USER_PERMISSIONS, CacheKeyFactory.id(userId),
                UserPermissionVO.class, () -> rbacMapper.findUserPermissions(userId));
    }

    // ── 内部校验 ───────────────────────────────────────────────

    private void requireUser(Long userId) {
        if (!userRepository.existsById(userId)) {
            throw new RuntimeException("用户不存在，ID: " + userId);
        }
    }

    private Role requireRole(Long roleId) {
        return roleRepository.findById(roleId)
                .orElseThrow(() -> new RuntimeException("角色不存在，ID: " + roleId));
    }

    private Permission requirePermission(Long permissionId) {
        return permissionRepository.findById(permissionId)
                .orElseThrow(() -> new RuntimeException("权限不存在，ID: " + permissionId));
    }
}